src/device-manager/cocoa/NLGenericTraitUpdatableDataSink.mm

```diff
@@ -0,0 +1,919 @@
+    // context information we need.
+    GenericTraitUpdatableDataSinkCompletionBlock _mCompletionHandler;
+    GenericTraitUpdatableDataSinkFailureBlock _mFailureHandler;
+    NSString * _mRequestName;
+}
+
+- (NSString *)GetCurrentRequest;
+@end
+
+@implementation NLGenericTraitUpdatableDataSink
+@synthesize resultCallbackQueue = _mAppCallbackQueue;
+
+/**
+ @note
+ This function can only be called by the ARC runtime
+ */
+- (void)dealloc
+{
+    // This method can only be called by ARC
+    // Let's not rely on this unpredictable mechanism for de-initialization
+    // application shall call ShutdownStack if it want to cleanly destroy everything before application termination
+    WDM_LOG_METHOD_SIG();
+
+    [self markTransactionCompleted];
+
+    _mRequestName = @"dealloc-Shutdown";
+
+    // we need to force the queue to be Weave work queue, as dealloc could be
+    // called at random queues (most probably from UI, when the app de-reference this GenericTraitUpdatableDataSink)
+    dispatch_sync(_mWeaveWorkQueue, ^() {
+        [self shutdown_Internal];
+    });
+}
+
+- (instancetype)init:(NSString *)name
+      weaveWorkQueue:(dispatch_queue_t)weaveWorkQueue
+    appCallbackQueue:(dispatch_queue_t)appCallbackQueue
+    genericTraitUpdatableDataSinkPtr: (nl::Weave::DeviceManager::GenericTraitUpdatableDataSink *)dataSinkPtr
+         nlWdmClient:(NLWdmClient *)nlWdmClient
+{
+    WEAVE_ERROR err = WEAVE_NO_ERROR;
+
+    WDM_LOG_METHOD_SIG();
+
+    self = [super init];
+    VerifyOrExit(nil != self, err = WEAVE_ERROR_NO_MEMORY);
+
+    _mWeaveWorkQueue = weaveWorkQueue;
+    _mAppCallbackQueue = appCallbackQueue;
+
+    _name = name;
+    _mWeaveCppGenericTraitUpdatableDataSink = dataSinkPtr;
+    _mNLWdmClient = nlWdmClient;
+
+    // simple bridge shall not increase our reference count
+    _mWeaveCppGenericTraitUpdatableDataSink->mpAppState = (__bridge void *) self;
+
+    [self markTransactionCompleted];
+
+exit:
+    id result = nil;
+    if (WEAVE_NO_ERROR == err) {
+        result = self;
+    } else {
+        WDM_LOG_ERROR(@"Error in init : (%d) %@\n", err, [NSString stringWithUTF8String:nl::ErrorStr(err)]);
+        [self shutdown_Internal];
+    }
+
+    return result;
+}
+
+static void handleGenericUpdatableDataSinkComplete(void * dataSink, void * reqState)
+{
+    WDM_LOG_DEBUG(@"handleGenericUpdatableDataSinkComplete");
+
+    NLGenericTraitUpdatableDataSink * sink = (__bridge NLGenericTraitUpdatableDataSink *) reqState;
+    [sink DispatchAsyncCompletionBlock:nil];
+}
+
+static void handleGenericUpdatableDataSinkError(void * dataSink, void * appReqState, WEAVE_ERROR code,
+    nl::Weave::DeviceManager::DeviceStatus * devStatus)
+{
+    WDM_LOG_DEBUG(@"handleGenericUpdatableDataSinkError");
+
+    NSError * error = nil;
+    NSDictionary * userInfo = nil;
+
+    NLGenericTraitUpdatableDataSink * nlDataSink = (__bridge NLGenericTraitUpdatableDataSink *) appReqState;
+    // ignore the pointer to C++ device manager
```

Comment:
typo -- the pointer is not a device manager pointer.

Reply:
Done, thanks